net/rhp/contracts_test.go

```diff
@@ -0,0 +1,603 @@
+		RenterOutput:    outputValue(types.Siacoins(10)),
+		HostPublicKey:   hostPubkey,
+		RenterPublicKey: renterPubkey,
+	}
+
+	if err := ValidateContractFormation(fc, currentHeight, settings); err != nil {
+		t.Fatal(err)
+	}
+
+	formChanges := []struct {
+		change func(fc types.FileContract) types.FileContract
+		error  string
+	}{
+		{
+			func(fc types.FileContract) types.FileContract {
+				fc.Filesize = 5
+				return fc
+			},
+			"initial filesize should be 0",
+		},
+		{
+			func(fc types.FileContract) types.FileContract {
+				fc.RevisionNumber = 5
+				return fc
+			},
+			"initial revision number should be 0",
+		},
+		{
+			func(fc types.FileContract) types.FileContract {
+				fc.FileMerkleRoot = types.Hash256{31: 1}
+				return fc
+			},
+			"initial Merkle root should be empty",
+		},
+		{
+			func(fc types.FileContract) types.FileContract {
+				fc.WindowStart = windowStart / 2
+				return fc
+			},
+			"contract ends too soon to safely submit the contract transaction",
+		},
+		{
+			func(fc types.FileContract) types.FileContract {
+				fc.WindowStart = currentHeight + 2*settings.MaxDuration
+				return fc
+			},
+			"contract duration is too long",
+		},
+		{
+			func(fc types.FileContract) types.FileContract {
+				fc.WindowEnd = windowStart + settings.WindowSize/2
+				return fc
+			},
+			"proof window is too small",
+		},
+		{
+			func(fc types.FileContract) types.FileContract {
+				fc.HostOutput.Address = types.Address{31: 1}
+				return fc
+			},
+			"wrong address for host valid output",
+		},
+		{
+			func(fc types.FileContract) types.FileContract {
+				fc.MissedHostValue = fc.HostOutput.Value.Mul64(2)
+				return fc
+			},
+			"host valid output value does not equal missed value",
+		},
+		{
+			func(fc types.FileContract) types.FileContract {
+				fc.HostOutput.Value = fc.HostOutput.Value.Mul64(1000)
+				fc.MissedHostValue = fc.HostOutput.Value
+				return fc
+			},
+			"wrong initial host output value",
+		},
+		{
+			func(fc types.FileContract) types.FileContract {
+				fc.TotalCollateral = settings.MaxCollateral.Mul64(1000)
+				fc.HostOutput.Value = settings.ContractFee.Add(fc.TotalCollateral)
+				fc.MissedHostValue = fc.HostOutput.Value
+				return fc
+			},
+			"excessive initial collateral",
+		},
+	}
+
+	for _, change := range formChanges {
+		changed := change.change(fc)
+		if err := ValidateContractFormation(changed, currentHeight, settings); err == nil {
+			t.Fatalf("expected error: %s", change.error)
+		}
+	}
+
+	currentHeight = windowEnd
+
+	renewal := fc
+	renewal.WindowEnd += 3 * settings.WindowSize
+	renewal.WindowStart = fc.WindowEnd + settings.WindowSize
+
+	if err := ValidateContractRenewal(fc, renewal, currentHeight, settings); err != nil {
+		t.Fatal(err)
+	}
+
+	renewalChanges := []struct {
+		change func(existing, renewal types.FileContract) types.FileContract
+		error  string
+	}{
+		{
+			func(existing, renewal types.FileContract) types.FileContract {
+				renewal.HostPublicKey[0] ^= 255
```

Comment:
careful -- if `types.FileContract` contained a slice, and you mutated that, then all subsequent tests would fail with the same error.

you *should* be fine here, but consider making a copy of `renewal` for each test case and then passing a pointer to the `change` function. You might also want to compare the exact error string; this is usually an anti-pattern, but frankly error strings don't change very often and this would ensure that we're hitting the errors we want to hit.

lastly, let's rename the struct fields to be consistent with https://github.com/SiaFoundation/core/blob/master/consensus/validation_test.go#L402